File: games/towerdefense/pathfinding.py
import heapq
from typing import List, Tuple, Optional

def heuristic(a: Tuple[int, int], b: Tuple[int, int]) -> int:
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

def astar(grid: List[List[int]], start: Tuple[int, int], end: Tuple[int, int]) -> Optional[List[Tuple[int, int]]]:
    rows, cols = len(grid), len(grid[0])
    if not (0 <= start[0] < rows and 0 <= start[1] < cols):
        return None
    if not (0 <= end[0] < rows and 0 <= end[1] < cols):
        return None
    if grid[start[0]][start[1]] == 1 or grid[end[0]][end[1]] == 1:
        return None
    if start == end:
        return [start]

    open_set = [(0, start)]
    came_from = {}
    g_score = {start: 0}

    while open_set:
        _, current = heapq.heappop(open_set)
        if current == end:
            path = []
            while current in came_from:
                path.append(current)
                current = came_from[current]
            path.append(start)
            return path[::-1]

        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            neighbor = (current[0] + dr, current[1] + dc)
            if 0 <= neighbor[0] < rows and 0 <= neighbor[1] < cols:
                if grid[neighbor[0]][neighbor[1]] == 1:
                    continue
                tentative_g = g_score[current] + 1
                if tentative_g < g_score.get(neighbor, float('inf')):
                    came_from[neighbor] = current
                    g_score[neighbor] = tentative_g
                    f_score = tentative_g + heuristic(neighbor, end)
                    heapq.heappush(open_set, (f_score, neighbor))

    return None

File: games/towerdefense/static/pathfinding.js
export function astar(grid, start, end) {
    const rows = grid.length, cols = grid[0].length;
    if (start[0] < 0 || start[0] >= rows || start[1] < 0 || start[1] >= cols) return null;
    if (end[0] < 0 || end[0] >= rows || end[1] < 0 || end[1] >= cols) return null;
    if (grid[start[0]][start[1]] === 1 || grid[end[0]][end[1]] === 1) return null;
    if (start[0] === end[0] && start[1] === end[1]) return [start];

    const openSet = [[0, start[0], start[1]]];
    const cameFrom = {};
    const gScore = {};
    gScore[`${start[0]},${start[1]}`] = 0;
    const dirs = [[-1,0],[1,0],[0,-1],[0,1]];

    while (openSet.length > 0) {
        openSet.sort((a, b) => a[0] - b[0]);
        const [, cr, cc] = openSet.shift();
        if (cr === end[0] && cc === end[1]) {
            const path = [];
            let key = `${cr},${cc}`;
            while (key) {
                const [r, c] = key.split(',').map(Number);
                path.unshift([r, c]);
                key = cameFrom[key];
            }
            return path;
        }
        for (const [dr, dc] of dirs) {
            const nr = cr + dr, nc = cc + dc;
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] !== 1) {
                const tentG = gScore[`${cr},${cc}`] + 1;
                const nKey = `${nr},${nc}`;
                if (tentG < (gScore[nKey] ?? Infinity)) {
                    cameFrom[nKey] = `${cr},${cc}`;
                    gScore[nKey] = tentG;
                    openSet.push([tentG + Math.abs(nr - end[0]) + Math.abs(nc - end[1]), nr, nc]);
                }
            }
        }
    }
    return null;
}
